File: php/src/gadgets/servlet/GadgetRenderingServlet.php
<?php
/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

require 'src/common/HttpServlet.php';
require 'src/gadgets/GadgetContext.php';
require 'src/gadgets/GadgetServer.php';
require 'src/common/RemoteContentRequest.php';
require 'src/common/RemoteContent.php';
require 'src/common/Cache.php';
require 'src/common/RemoteContentFetcher.php';
require 'src/gadgets/GadgetSpecParser.php';
require 'src/gadgets/Gadget.php';
require 'src/gadgets/GadgetId.php';
require 'src/gadgets/UserPrefs.php';
require 'src/gadgets/Substitutions.php';
require 'src/gadgets/LocaleSpec.php';
require 'src/gadgets/LocaleMessageBundle.php';
require 'src/gadgets/GadgetBlacklist.php';
require 'src/common/Locale.php';
require 'src/gadgets/UserPref.php';
require 'src/gadgets/ViewSpec.php';
require 'src/gadgets/FeatureSpec.php';
require 'src/gadgets/MessageBundleParser.php';
require 'src/gadgets/MessageBundle.php';
require 'src/gadgets/GadgetFeatureRegistry.php';
require 'src/gadgets/GadgetFeatureFactory.php';
require 'src/gadgets/GadgetFeature.php';
require 'src/gadgets/JsLibraryFeatureFactory.php';
require 'src/gadgets/JsLibrary.php';
require 'src/gadgets/HttpUtil.php';
require 'src/gadgets/ContainerConfig.php';
require 'src/common/JsMin.php';
require 'src/common/SecurityTokenDecoder.php';
require 'src/common/SecurityToken.php';
require 'src/common/BlobCrypter.php';
require 'src/gadgets/rewrite/ContentRewriter.php';
require 'src/gadgets/rewrite/ContentRewriteFeature.php';

/**
 * This class deals with the gadget rendering requests (in default config this
 * would be /gadgets/ifr?url=<some gadget's url>). It uses the gadget server and
 * gadget context to render the xml to a valid html file, and outputs it.
 *
 */
class GadgetRenderingServlet extends HttpServlet {
  private $context;

  /**
   * Creates the gadget using the GadgetServer class and calls outputGadget
   *
   */
  public function doGet() {
    try {
      if (empty($_GET['url'])) {
        throw new GadgetException("Missing required parameter: url");
      }
      // GadgetContext builds up all the contextual variables (based on the url or post)
      // plus instances all required classes (feature registry, fetcher, blacklist, etc)
      $this->context = new GadgetContext('GADGET');
      // Unfortunatly we can't do caja content filtering here, hoping we'll have a RPC service
      // or command line caja to use for this at some point
      $gadgetServer = new GadgetServer();
      $gadget = $gadgetServer->processGadget($this->context);
      $this->outputGadget($gadget, $this->context);
    } catch (Exception $e) {
      $this->outputError($e);
    }
  }

  /**
   * If an error occured (Exception) this function echo's the Exception's message
   * and if the config['debug'] is true, shows the debug backtrace in a div
   *
   * @param Exception $e the exception to show
   */
  private function outputError($e) {
    header("HTTP/1.0 400 Bad Request", true, 400);
    echo "<html><body>";
    echo "<h1>Error</h1>";
    echo $e->getMessage();
    if (Config::get('debug')) {
      echo "<p><b>Debug backtrace</b></p><div style='overflow:auto; height:300px; border:1px solid #000000'><pre>";
      print_r(debug_backtrace());
      echo "</pre></div>>";
    }
    echo "</body></html>";
  }

  /**
   * Takes the gadget to output, and depending on its content type calls either outputHtml-
   * or outputUrlGadget
   *
   * @param Gadget $gadget gadget to render
   * @param string $view the view to render (only valid with a html content type)
   */
  private function outputGadget($gadget, $context) {
    $view = HttpUtil::getView($gadget, $context);
    switch ($view->getType()) {
      case 'HTML':
        $this->outputHtmlGadget($gadget, $context, $view);
        break;
      case 'URL':
        $this->outputUrlGadget($gadget, $context, $view);
        break;
    }
  }

  /**
   * Outputs a html content type gadget.
   * It creates a html page, with the javascripts from the features inline into the page, plus
   * calls to 'gadgets.config.init' with the container configuration (config/container.js) and
   * 'gadgets.Prefs.setMessages_' with all the substitutions. For external javascripts it adds
   * a <script> tag.
   *
   * @param Gadget $gadget
   * @param GadgetContext $context
   */
  private function outputHtmlGadget($gadget, $context, $view) {
    $content = '';
    $externJs = '';
    $externFmt = "<script src=\"%s\"></script>";
    $forcedLibs = $context->getForcedJsLibs();
    // allow the &libs=.. param to override our forced js libs configuration value
    if (empty($forcedLibs)) {
      $forcedLibs = Config::get('focedJsLibs');
    }
    $this->setContentType("text/html; charset=UTF-8");
    if ($context->getIgnoreCache()) {
      // no cache was requested, set non-caching-headers
      $this->setNoCache(true);
    } elseif (isset($_GET['v'])) {
      // version was given, cache for a long long time (a year)
      $this->setCacheTime(365 * 24 * 60 * 60);
    } else {
      // no version was given, cache for 5 minutes
      $this->setCacheTime(5 * 60);
    }
    // Was a privacy policy header configured? if so set it
    if (Config::get('P3P') != '') {
      header("P3P: " . Config::get('P3P'));
    }
    if (! $view->getQuirks()) {
      $content .= "<!DOCTYPE HTML PUBLIC \"-//W3C//DTD HTML 4.01//EN\" \"http://www.w3.org/TR/html4/strict.dtd\">\n";
    }
    $content .= "<html><head><meta http-equiv=\"Content-Type\" content=\"text/html; charset=utf-8\"/><style type=\"text/css\">" . Config::get('gadget_css') . "</style></head><body>\n";
    // Forced libs first.
    if (! empty($forcedLibs)) {
      $libs = explode(':', $forcedLibs);
      $content .= sprintf($externFmt, Config::get('default_js_prefix') . $this->getJsUrl($libs, $gadget) . "&container=" . $context->getContainer()) . "\n";
    }
    $content .= "<script>\n";

    if (! empty($forcedLibs)) {
      // if some of the feature libraries are externalized (through a browser cachable <script src="/gadgets/js/opensocial-0.7:settitle.js">
      // type url), then we don't want to include dependencies twice, so find the complete features chain, so we can skip over those
      $forcedLibsArray = explode(':', $forcedLibs);
      $registry = $this->context->getRegistry();
      $missing = array();
      $registry->getIncludedFeatures($forcedLibsArray, $forcedLibsArray, $missing);
    }
    foreach ($gadget->getJsLibraries() as $library) {
      $type = $library->getType();
      if ($type == 'URL') {
        // TODO: This case needs to be handled more gracefully by the js
        // servlet. We should probably inline external JS as well.
        $externJs .= sprintf($externFmt, $library->getContent()) . "\n";
        // else check if there are no forcedLibs, or if it wasn't included in their dep chain
      } elseif (empty($forcedLibs) || ! in_array($library->getFeatureName(), $forcedLibsArray)) {
        $content .= $library->getContent();
      }
      // otherwise it was already included by config.forceJsLibs.
    }
    $content .= $this->appendJsConfig($context, $gadget, ! empty($forcedLibs)) . $this->appendMessages($gadget) . $this->appendPreferences($gadget) . $this->appendPreloads($gadget, $context) . "</script>";
    if (strlen($externJs) > 0) {
      $content .= $externJs;
    }
    $gadgetExceptions = array();
    $rewriter = new ContentRewriter();
    if ($rewriter->rewriteGadgetView($gadget, $view)) {
      $content .= $gadget->getSubstitutions()->substitute($view->getRewrittenContent());
    } else {
      $content .= $gadget->getSubstitutions()->substitute($view->getContent());
    }
    if (empty($content)) {
      // Unknown view
      $gadgetExceptions[] = "View: '" . $context->getView() . "' invalid for gadget: " . $gadget->getId()->getKey();
    }
    if (count($gadgetExceptions)) {
      throw new GadgetException(print_r($gadgetExceptions, true));
    }
    $content .= "\n<script>gadgets.util.runOnLoadHandlers();</script></body>\n</html>";
    echo $content;
  }

  /**
   * Output's a URL content type gadget, it adds libs=<list:of:js:libraries>.js and user preferences
   * to the href url, and redirects the browser to it
   *
   * @param Gadget $gadget
   */
  private function outputUrlGadget($gadget, $context, $view) {
    // Preserve existing query string parameters.
    $redirURI = $view->getHref();
    $queryStr = strpos($redirURI, '?') !== false ? substr($redirURI, strpos($redirURI, '?')) : '';
    $query = $queryStr;
    // TODO: userprefs on the fragment rather than query string
    $query .= $this->getPrefsQueryString($gadget->getUserPrefValues());
    $libs = array();
    $forcedLibs = Config::get('focedJsLibs');
    if ($forcedLibs == null) {
      $reqs = $gadget->getRequires();
      foreach ($reqs as $key => $val) {
        $libs[] = $key;
      }
    } else {
      $libs = explode(':', $forcedLibs);
    }
    $query .= $this->appendLibsToQuery($libs, $gadget);
    // code bugs out with me because of the invalid url syntax since we dont have a URI class to fix it for us
    // this works around that
    if (substr($query, 0, 1) == '&') {
      $query = '?' . substr($query, 1);
    }
    $redirURI .= $query;
    header('Location: ' . $redirURI);
    die();
  }

  /**
   * Returns the requested libs (from getjsUrl) with the libs_param_name prepended
   * ie: in libs=core:caja:etc.js format
   *
   * @param string $libs the libraries
   * @param Gadget $gadget
   * @return string the libs=... string to append to the redirection url
   */
  private function appendLibsToQuery($libs, $gadget) {
    $ret = "&";
    $ret .= Config::get('libs_param_name');
    $ret .= "=";
    $ret .= $this->getJsUrl($libs, $gadget);
    return $ret;
  }

  /**
   * Returns the user preferences in &up_<name>=<val> format
   *
   * @param array $libs array of features this gadget requires
   * @param Gadget $gadget
   * @return string the up_<name>=<val> string to use in the redirection url
   */
  private function getPrefsQueryString($prefVals) {
    $ret = '';
    foreach ($prefVals->getPrefs() as $key => $val) {
      $ret .= '&';
      $ret .= Config::get('userpref_param_prefix');
      $ret .= urlencode($key);
      $ret .= '=';
      $ret .= urlencode($val);
    }
    return $ret;
  }

  /**
   * generates the library string (core:caja:etc.js) including a checksum of all the
   * javascript content (?v=<md5 of js>) for cache busting
   *
   * @param string $libs
   * @param Gadget $gadget
   * @return string the list of libraries in core:caja:etc.js?v=checksum> format
   */
  private function getJsUrl($libs, $gadget) {
    $buf = '';
    if (! is_array($libs) || ! count($libs)) {
      $buf = 'core';
    } else {
      $firstDone = false;
      foreach ($libs as $lib) {
        if ($firstDone) {
          $buf .= ':';
        } else {
          $firstDone = true;
        }
        $buf .= $lib;
      }
    }
    $cache = Config::get('feature_cache');
    $cache = new $cache();
    if (($md5 = $cache->get(md5('getJsUrlMD5'))) === false) {
      $registry = $this->context->getRegistry();
      $features = $registry->getAllFeatures();
      // Build a version string from the md5() checksum of all included javascript
      // to ensure the client always has the right version
      $inlineJs = '';
      foreach ($features as $feature) {
        $library = $feature->getFeature();
        $libs = $library->getLibraries($this->context->getRenderingContext());
        foreach ($libs as $lib) {
          $inlineJs .= $lib->getContent();
        }
      }
      $md5 = md5($inlineJs);
      $cache->set(md5('getJsUrlMD5'), $md5);
    }
    $buf .= ".js?v=" . $md5;
    return $buf;
  }

  private function appendJsConfig($context, $gadget, $hasForcedLibs) {
    $container = $context->getContainer();
    $containerConfig = $context->getContainerConfig();
    //TODO some day we should parse the forcedLibs too, and include their config selectivly as well
    // for now we just include everything if forced libs is set.
    if ($hasForcedLibs) {
      $gadgetConfig = $containerConfig->getConfig($container, 'gadgets.features');
    } else {
      $gadgetConfig = array();
      $featureConfig = $containerConfig->getConfig($container, 'gadgets.features');
      foreach ($gadget->getJsLibraries() as $library) {
        $feature = $library->getFeatureName();
        if (! isset($gadgetConfig[$feature]) && ! empty($featureConfig[$feature])) {
          $gadgetConfig[$feature] = $featureConfig[$feature];
        }
      }
    }

    // Add gadgets.util support. This is calculated dynamically based on request inputs.
    // See java/org/apache/shindig/gadgets/render/RenderingContentRewriter.java for reference.
    $requires = array();
    foreach ($gadget->getRequires() as $feature) {
      $requires[$feature->name] = new EmptyClass();
    }
    $gadgetConfig['core.util'] = $requires;

    return "gadgets.config.init(" . json_encode($gadgetConfig) . ");\n";
  }

  private function appendMessages($gadget) {
    $msgs = '';
    if ($gadget->getMessageBundle()) {
      $bundle = $gadget->getMessageBundle();
      $msgs = json_encode($bundle->getMessages());
    }
    return "gadgets.Prefs.setMessages_($msgs);\n";
  }

  private function appendPreferences(Gadget $gadget) {
    $prefs = json_encode($gadget->getUserPrefValues()->getPrefs());
    return "gadgets.Prefs.setDefaultPrefs_($prefs);\n";
  }

  /**
   * Appends data from <Preload> elements to make them available to
   * gadgets.io.
   *
   * @param gadget
   */
  private function appendPreloads(Gadget $gadget, GadgetContext $context) {
    $resp = Array();
    $gadgetSigner = Config::get('security_token_signer');
    $gadgetSigner = new $gadgetSigner();
    $token = '';
    try {
      $token = $context->extractAndValidateToken($gadgetSigner);
    } catch (Exception $e) {
      $token = '';
      // no token given, safe to ignore
    }
    $unsignedRequests = $unsignedContexts = Array();
    $signedRequests = Array();
    foreach ($gadget->getPreloads() as $preload) {
      try {
        if (($preload->getAuth() == Auth::$NONE || $token != null) && (count($preload->getViews()) == 0 || in_array($context->getView(), $preload->getViews()))) {
          $request = new RemoteContentRequest($preload->getHref());
          $request->createRemoteContentRequestWithUri($preload->getHref());
          $request->getOptions()->ownerSigned = $preload->isSignOwner();
          $request->getOptions()->viewerSigned = $preload->isSignViewer();
          switch (strtoupper(trim($preload->getAuth()))) {
            case "NONE":
              //						Unify all unsigned requests to one single multi request
              $unsignedRequests[] = $request;
              $unsignedContexts[] = $context;
              break;
            case "SIGNED":
              //						Unify all signed requests to one single multi request
              $signingFetcherFactory = new SigningFetcherFactory(Config::get("private_key_file"));
              $fetcher = $signingFetcherFactory->getSigningFetcher(new BasicRemoteContentFetcher(), $token);
              $req = $fetcher->signRequest($preload->getHref(), $request->getMethod());
              $req->setNotSignedUri($preload->getHref());
              $signedRequests[] = $req;
              break;
            default:
              @ob_end_clean();
              header("HTTP/1.0 500 Internal Server Error", true);
              echo "<html><body><h1>" . "500 - Internal Server Error" . "</h1></body></html>";
              die();
          }
        }
      } catch (Exception $e) {
        throw new Exception($e);
      }
    }
    if (count($unsignedRequests)) {
      try {
        $brc = new BasicRemoteContent();
        $responses = $brc->multiFetch($unsignedRequests, $unsignedContexts);
        foreach ($responses as $response) {
          $resp[$response->getUrl()] = array(
              'body' => $response->getResponseContent(),
              'rc' => $response->getHttpCode());
        }
      } catch (Exception $e) {
        throw new Exception($e);
      }
    }
    if (count($signedRequests)) {
      try {
        $fetcher = $signingFetcherFactory->getSigningFetcher(new BasicRemoteContentFetcher(), $token);
        $responses = $fetcher->multiFetchRequest($signedRequests);
        foreach ($responses as $response) {
          $resp[$response->getNotSignedUrl()] = array(
              'body' => $response->getResponseContent(),
              'rc' => $response->getHttpCode());
        }
      } catch (Exception $e) {
        throw new Exception($e);
      }
    }
    $resp = count($resp) ? json_encode($resp) : "{}";
    return "gadgets.io.preloaded_ = " . $resp . ";\n";
  }
}

/*
 * An empty class for generate "{}" code in JavaScript.
 */
class EmptyClass {
}

File: trunk/java/gadgets/src/main/java/org/apache/shindig/gadgets/servlet/GadgetRenderingServlet.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package org.apache.shindig.gadgets.servlet;

import org.apache.shindig.common.logging.i18n.MessageKeys;
import org.apache.shindig.common.servlet.HttpUtil;
import org.apache.shindig.common.servlet.InjectedServlet;
import org.apache.shindig.common.uri.UriBuilder;
import org.apache.shindig.gadgets.GadgetContext;
import org.apache.shindig.gadgets.http.HttpRequest;
import org.apache.shindig.gadgets.render.Renderer;
import org.apache.shindig.gadgets.render.RenderingResults;
import org.apache.shindig.gadgets.uri.IframeUriManager;
import org.apache.shindig.gadgets.uri.UriStatus;
import org.apache.shindig.gadgets.uri.UriCommon.Param;

import com.google.common.base.Strings;
import com.google.inject.Inject;

import org.apache.commons.lang3.StringEscapeUtils;

import java.io.IOException;
import java.util.logging.Level;
import java.util.logging.Logger;

import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;

/**
 * Servlet for rendering Gadgets.
 */
public class GadgetRenderingServlet extends InjectedServlet {

  private static final long serialVersionUID = -5634040113214794888L;

  static final int DEFAULT_CACHE_TTL = 60 * 5;

  //class name for logging purpose
  private static final String classname = GadgetRenderingServlet.class.getName();
  private static final Logger LOG = Logger.getLogger(classname,MessageKeys.MESSAGES);

  protected transient Renderer renderer;
  protected transient IframeUriManager iframeUriManager;

  @Inject
  public void setRenderer(Renderer renderer) {
    checkInitialized();
    this.renderer = renderer;
  }

  @Inject
  public void setIframeUriManager(IframeUriManager iframeUriManager) {
    checkInitialized();
    this.iframeUriManager = iframeUriManager;
  }

  @Override
  protected void doGet(HttpServletRequest req, HttpServletResponse resp) throws IOException {
    // If an If-Modified-Since header is ever provided, we always say
    // not modified. This is because when there actually is a change,
    // cache busting should occur.
    UriStatus urlStatus = getUrlStatus(req);
    if (req.getHeader("If-Modified-Since") != null &&
        !"1".equals(req.getParameter("nocache")) &&
        urlStatus == UriStatus.VALID_VERSIONED) {
      resp.setStatus(HttpServletResponse.SC_NOT_MODIFIED);
      return;
    }
    render(req, resp, urlStatus);
  }

  @Override
  protected void doPost(HttpServletRequest req, HttpServletResponse resp) throws IOException {
    render(req, resp, getUrlStatus(req));
  }

  private void render(HttpServletRequest req, HttpServletResponse resp, UriStatus urlstatus)
      throws IOException {
    if (req.getHeader(HttpRequest.DOS_PREVENTION_HEADER) != null) {
      // Refuse to render for any request that came from us.
      // TODO: Is this necessary for any other type of request? Rendering seems to be the only one
      // that can potentially result in an infinite loop.
      resp.sendError(HttpServletResponse.SC_FORBIDDEN);
      return;
    }

    resp.setContentType("text/html");
    resp.setCharacterEncoding("UTF-8");

    GadgetContext context = new HttpGadgetContext(req);
    RenderingResults results = renderer.render(context);

    // process the rendering results
    postGadgetRendering(new PostGadgetRenderingParams(req, resp, urlstatus, context, results));
  }

  /**
   * Implementations that extend this class are strongly discouraged from overriding this method.
   * To customize the behavior please override the hook methods for each of the
   * RenderingResults.Status enum values instead.
   */
  protected void postGadgetRendering(PostGadgetRenderingParams params) throws IOException {
    switch (params.getResults().getStatus()) {
      case OK:
        onOkRenderingResultsStatus(params);
        break;
      case ERROR:
        onErrorRenderingResultsStatus(params);
        break;
      case MUST_REDIRECT:
        onMustRedirectRenderingResultsStatus(params);
        break;
    }
  }

  protected void onOkRenderingResultsStatus(PostGadgetRenderingParams params)
      throws IOException {
    UriStatus urlStatus = params.getUrlStatus();
    HttpServletResponse resp = params.getResponse();
    if (params.getContext().getIgnoreCache() ||
        urlStatus == UriStatus.INVALID_VERSION) {
      HttpUtil.setCachingHeaders(resp, 0);
    } else if (urlStatus == UriStatus.VALID_VERSIONED) {
      // Versioned files get cached indefinitely
      HttpUtil.setCachingHeaders(resp, true);
    } else {
      // Unversioned files get cached for 5 minutes by default, but this can be overridden
      // with a query parameter.
      int ttl = DEFAULT_CACHE_TTL;
      String ttlStr = params.getRequest().getParameter(Param.REFRESH.getKey());
      if (!Strings.isNullOrEmpty(ttlStr)) {
        try {
          ttl = Integer.parseInt(ttlStr);
        } catch (NumberFormatException e) {
          // Ignore malformed TTL value
          if (LOG.isLoggable(Level.INFO)) {
            LOG.logp(Level.INFO, classname, "onOkRenderingResultsStatus", MessageKeys.MALFORMED_TTL_VALUE, new Object[] {ttlStr});
          }
        }
      }
      HttpUtil.setCachingHeaders(resp, ttl, true);
    }
    resp.getWriter().print(params.getResults().getContent());
  }

  protected void onErrorRenderingResultsStatus(PostGadgetRenderingParams params)
      throws IOException {
    HttpServletResponse resp = params.getResponse();
    resp.setStatus(params.getResults().getHttpStatusCode());
    resp.getWriter().print(StringEscapeUtils.escapeHtml4(params.getResults().getErrorMessage()));
  }

  protected void onMustRedirectRenderingResultsStatus(PostGadgetRenderingParams params)
      throws IOException {
     params.getResponse().sendRedirect(params.getResults().getRedirect().toString());
  }

  private UriStatus getUrlStatus(HttpServletRequest req) {
    return iframeUriManager.validateRenderingUri(new UriBuilder(req).toUri());
  }

  /**
   * Contains the input parameters for post rendering methods.
   */
  protected static class PostGadgetRenderingParams {
    private HttpServletRequest req;
    private HttpServletResponse resp;
    private UriStatus urlStatus;
    private GadgetContext context;
    private RenderingResults results;

    public PostGadgetRenderingParams (HttpServletRequest req, HttpServletResponse resp,
      UriStatus urlStatus, GadgetContext context, RenderingResults results) {
      this.req = req;
      this.resp = resp;
      this.urlStatus = urlStatus;
      this.context = context;
      this.results = results;
    }

    public HttpServletRequest getRequest() {
      return req;
    }

    public HttpServletResponse getResponse() {
      return resp;
    }

    public UriStatus getUrlStatus() {
      return urlStatus;
    }

    public GadgetContext getContext() {
      return context;
    }

    public RenderingResults getResults() {
      return results;
    }
  }
}
